trace/trace.go

```diff
@@ -157,11 +171,171 @@ func decodeHex(h string, b []byte) error {
 	return nil
 }
 
+// TraceState provides additional vendor-specific trace identification information
+// across different distributed tracing systems. It represents an immutable list consisting
+// of key/value pairs. There can be a maximum of 32 entries in the list.
+//
+// Key and value of each list member must be valid according to the W3C Trace Context specification
+// (see https://www.w3.org/TR/trace-context-1/#key and https://www.w3.org/TR/trace-context-1/#value
+// respectively).
+//
+// Trace state must be valid according to the W3C Trace Context specification at all times. All
+// mutating operations validate their input and, in case of valid parameters, return a new TraceState.
+type TraceState struct { //nolint:golint
+	kvs []label.KeyValue
+}
+
+var _ json.Marshaler = TraceState{}
+var keyFormatRegExp = regexp.MustCompile(
+	`^((` + traceStateKeyFormat + `)|(` + traceStateKeyFormatWithMultiTenantVendor + `))$`,
+)
+var valueFormatRegExp = regexp.MustCompile(`^(` + traceStateValueFormat + `)$`)
+
+// MarshalJSON implements a custom marshal function to encode trace state.
+func (ts TraceState) MarshalJSON() ([]byte, error) {
+	return json.Marshal(ts.kvs)
+}
+
+// String returns trace state as a string valid according to the
+// W3C Trace Context specification.
+func (ts TraceState) String() string {
```

Comment:
I was comparing with the source in `propagation/baggage.go` (`Baggage.Inject()`) and wonder if the differences are intentional. These in theory are the same format, but one is escaped. Should this use TrimSpace()?

Reply:
As `TraceState` can include only valid key-value entries according to [spec](https://www.w3.org/TR/trace-context/#list-members) (i.e. this is enforced via regexps), would escaping / trimming be required for representing `TraceState` as a string? 

My understanding is that for example `key1=valX&valY` is a valid member and I would expect the string representation to include 'unescaped' string instead of representing it as `valX%26valY` (as this can theoretically also be a valid entry according to the spec). I would not be able to tell what is the actual value of `key1`, whether the escaped or unescaped version. The same would go for leading / trailing space - unless these are allowed by the spec, they should not find their way into entries of a valid `TraceState`.